app/services/weather.py

```diff
@@ -0,0 +1,329 @@
+        self._ttl = max(0, int(ttl_seconds))
+        self._store: Dict[str, Tuple[float, Any]] = {}
+
+    def _key(self, city: str, start: str, end: str) -> str:
+        return json.dumps([city.strip().lower(), start, end])
+
+    def get(self, city: str, start: str, end: str) -> Optional[Any]:
+        if self._ttl == 0:
+            return None
+        key = self._key(city, start, end)
+        entry = self._store.get(key)
+        if not entry:
+            return None
+        ts, value = entry
+        if time.time() - ts > self._ttl:
+            self._store.pop(key, None)
+            return None
+        return value
+
+    def set(self, city: str, start: str, end: str, value: Any) -> None:
+        if self._ttl == 0:
+            return
+        key = self._key(city, start, end)
+        self._store[key] = (time.time(), value)
+
+
+# Process-wide cache instance (initialized on first WeatherService creation)
+_GLOBAL_CACHE: Optional[_TTLCache] = None
+
+
+def _coerce_date(d: Any) -> datetime:
+    if isinstance(d, datetime):
+        return d
+    if isinstance(d, (int, float)):
+        return datetime.fromtimestamp(d)
+    if isinstance(d, str):
+        # Accept common formats; fallback to ISO
+        for fmt in ("%Y-%m-%d", "%Y/%m/%d", "%d-%m-%Y", "%d/%m/%Y"):
+            try:
+                return datetime.strptime(d, fmt)
+            except ValueError:
+                pass
+        return datetime.fromisoformat(d)
+    raise ValueError(f"Unsupported date value: {d}")
+
+
+def _normalize_label(weather_main: str, temp_min: float, temp_max: float, precipitation_mm: float) -> str:
+    main_lower = (weather_main or "").strip().lower()
+    if "snow" in main_lower:
+        return "snowy"
+    if precipitation_mm >= 2.0 or "rain" in main_lower or "drizzle" in main_lower:
+        return "rainy"
+    if temp_max <= 0.0:
+        return "freezing"
+    if temp_max < 10.0:
+        return "cold"
+    if temp_min > 25.0:
+        return "hot"
+    if "cloud" in main_lower:
+        return "cloudy"
+    return "sunny"
+
+
+def _aggregate_to_daily(list_items: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
+    """Aggregate OpenWeather 3-hourly forecast to per-day buckets.
+
+    We compute min/max temps, sum precipitation, and choose the most frequent
+    weather "main"/description for the day.
+    """
+    buckets: Dict[str, Dict[str, Any]] = {}
+    for item in list_items:
+        dt_txt = item.get("dt_txt")  # e.g., "2025-09-20 12:00:00"
+        if not dt_txt:
+            # Fallback to unix timestamp
+            dt = datetime.fromtimestamp(int(item.get("dt", 0)))
+        else:
+            dt = datetime.strptime(dt_txt, "%Y-%m-%d %H:%M:%S")
+        day_key = dt.strftime("%Y-%m-%d")
+
+        main = item.get("main", {})
+        temp_min = float(main.get("temp_min", main.get("temp", 0.0)))
+        temp_max = float(main.get("temp_max", main.get("temp", 0.0)))
+
+        weather_arr = item.get("weather", []) or [{}]
```

Comment:
Only the first weather object is used for each time slot, which may miss important conditions. Aggregate all weather objects for a more accurate label.
```
weather_arr = item.get("weather", [])
for w in weather_arr:
    # count each 'main' and 'description'
```